openedx/core/djangoapps/course_groups/models.py

```diff
@@ -58,6 +59,96 @@ def create(cls, name, course_id, group_type=COHORT):
+    # For django 1.8 upgrade, just remove this method and allow the base method to be called instead.
+    # Reference: https://docs.djangoproject.com/en/1.6/ref/models/instances/, under "Validating Objects"
+    def full_clean(self, **kwargs):
+        self.clean_fields()
+        self.clean()
+        if 'validate_unique' not in kwargs or kwargs['validate_unique'] is True:
+            self.validate_unique()
+
+    def clean_fields(self, *args, **kwargs):
+        if self.course_id is None:
+            self.course_id = self.course_user_group.course_id
+        super(CohortMembership, self).clean_fields(*args, **kwargs)
+
+    def clean(self):
+        if self.course_user_group.group_type != CourseUserGroup.COHORT:  # pylint: disable=E1101
+            raise ValidationError("CohortMembership cannot be used with CourseGroup types other than COHORT")
+        if self.course_user_group.course_id != self.course_id:
+            raise ValidationError("Non-matching course_ids provided")
+
+    def save(self, *args, **kwargs):
+        self.full_clean(validate_unique=False)
+
+        max_retries = 5
+        for num_attempts in range(max_retries):
+            saved_membership = None
+            try:
+                # The following 2 "transaction" lines force a fresh read, they can be removed once we're on django 1.8
+                # http://stackoverflow.com/questions/3346124/how-do-i-force-django-to-ignore-any-caches-and-reload-data
+                with transaction.commit_manually():
+                    transaction.commit()
+
+                saved_membership = CohortMembership.objects.get(
+                    user__id=self.user.id,  # pylint: disable=E1101
+                    course_id=self.course_id
+                )
+            except CohortMembership.DoesNotExist:
+                pass
+
+            if saved_membership is not None:
+                if saved_membership.course_user_group == self.course_user_group:
+                    raise ValueError("User {user_name} already present in cohort {cohort_name}".format(
+                        user_name=self.user.username,  # pylint: disable=E1101
+                        cohort_name=self.course_user_group.name
+                    ))
+                self.previous_cohort = saved_membership.course_user_group
+                self.previous_cohort_name = saved_membership.course_user_group.name
+                self.previous_cohort_id = saved_membership.course_user_group.id
+
+            try:
+                with transaction.commit_on_success():
+                    self.course_user_group.users.add(self.user)  # pylint: disable=E1101
+                    self.user.course_groups.add(self.course_user_group)
+                    if saved_membership is not None:
+                        self.previous_cohort.users.remove(self.user)
+                        updated = CohortMembership.objects.filter(
```

Comment:
Why not use a `select_for_update()` here?:

https://docs.djangoproject.com/en/1.4/ref/models/querysets/#select-for-update

If you're certain that a single `UPDATE..WHERE` SQL statement is generated by the `filter().update()`, then it seems this code will work. But `select_for_update()` seems designed for the situation where the code wants to read a row/rows and update them within a transaction. (Just wondering...)